pkg/cmd/pulumi/templatecmd/template_publish.go

```diff
@@ -0,0 +1,186 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package templatecmd
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/blang/semver"
+	"github.com/pulumi/pulumi/pkg/v3/backend"
+	"github.com/pulumi/pulumi/pkg/v3/backend/display"
+	cmdBackend "github.com/pulumi/pulumi/pkg/v3/cmd/pulumi/backend"
+	pkgWorkspace "github.com/pulumi/pulumi/pkg/v3/workspace"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/archive"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+	"github.com/spf13/cobra"
+)
+
+type publishTemplateArgs struct {
+	publisher string
+	version   string
+	name      string
+}
+
+type templatePublishCmd struct {
+	defaultOrg func(context.Context, backend.Backend, *workspace.Project) (string, error)
+}
+
+func newTemplatePublishCmd() *cobra.Command {
+	args := publishTemplateArgs{}
+	var tplPublishCmd templatePublishCmd
+
+	cmd := &cobra.Command{
+		Use:   "publish <directory>",
+		Args:  cmdutil.ExactArgs(1),
+		Short: "Publish a template to the Pulumi Registry",
+		Long: "Publish a template to the Pulumi Registry.\n\n" +
+			"This command publishes a template directory to the Pulumi Registry. " +
+			"The directory will be archived as a .tar.gz file and uploaded.",
+		RunE: func(cmd *cobra.Command, cliArgs []string) error {
+			ctx := cmd.Context()
+			tplPublishCmd.defaultOrg = backend.GetDefaultOrg
+			return tplPublishCmd.Run(ctx, args, cliArgs[0])
+		},
+	}
+
+	cmd.Flags().StringVar(
+		&args.version, "version", "",
+		"The version of the template (required, semver format)")
+	cmd.Flags().StringVar(
+		&args.name, "name", "",
+		"The name of the template (required)")
+	cmd.Flags().StringVar(
+		&args.publisher, "publisher", "",
+		"The publisher of the template (e.g., 'pulumi'). Defaults to the default organization in your pulumi config.")
+	err := cmd.MarkFlagRequired("version")
+	if err != nil {
+		panic(err)
+	}
+	err = cmd.MarkFlagRequired("name")
+	if err != nil {
+		panic(err)
+	}
+
+	return cmd
+}
+
+func (cmd *templatePublishCmd) Run(
+	ctx context.Context,
+	args publishTemplateArgs,
+	templateDir string,
+) error {
+	if _, err := os.Stat(templateDir); os.IsNotExist(err) {
+		return fmt.Errorf("template directory does not exist: %s", templateDir)
+	}
+
+	if args.version == "" {
+		return errors.New("version is required, please specify --version")
+	}
+	version, err := semver.ParseTolerant(args.version)
+	if err != nil {
+		return fmt.Errorf("invalid version format: %w", err)
+	}
```

Comment:
Optional style nit, I would move arg parsing to the enclosing function, and switch the type of `args.version` from `string` to `semver.Version`.

Reply:
Much of a muchness, new style would be testing the newTemplatePublishCmd directly rather than a cmd struct. But we can rejig to that style once this is merged.